main/src/ca/uwaterloo/flix/language/phase/Redundancy.scala

```diff
@@ -920,4 +1001,21 @@ object Redundancy extends Phase[TypedAst.Root, TypedAst.Root] {
     }
   }
 
+  /**
+   * @param defn def whose context we are in
+   * @param params number of parameters required to execute recursion
+   * @param immediate true iff immediately after the def, or within a nested lambda immediately after the def
+   */
+  private case class RecursionContext(defn: Symbol.DefnSym, params: Int, immediate: Boolean, applies: Int) {
+    def addParam: RecursionContext = copy(params = params + 1)
+    def resetApplies: RecursionContext = copy(applies = 0)
+    def addApply: RecursionContext = copy(applies = applies + 1)
+    def stopBuilding: RecursionContext = copy(immediate = false)
+  }
+
+  private object RecursionContext {
+    def of(defn: Def): RecursionContext = {
+      RecursionContext(defn.sym, defn.fparams.size, immediate = true, applies = 0)
+    }
```

Comment:
I am intuitively trying to understand the algorithm, but I am not quite there yet. But it is starting to make sense.

The defn and params makes sense to me, although "arity" might be a better name for params.

Not sure what immediate is for.

applies counts the number of current applies.

Thus if you see an apply you increment it, and if you then later see a def where the number of applies = arity you have found a recursive call. Then you need to make sure this happens on all branches for there to be an error. Correct?

Reply:
Yes, that's correct. (I should have done a bit more documenting!) "Immediate" is for ensuring we're only counting lambdas as long as we're directly under the definition, since a lambda elsewhere in the tree shouldn't affect the arity of the function.